### PR TITLE
Upload largest icon from ipa to S3

### PR DESCRIPTION
**Pull request by piotrdubiel**
I have added largest icon extraction from .ipa and upload to S3. Do you think largest icon is enough or should it extract all icons?

### CODE REVIEW

**Comment by ohwutup** [comment]
Hey @piotrdubiel, thanks for the PR! 

I am not totally sure what the goal of this PR is however, do you think you could write up a little about the use case or reference an issue where this has been discussed? 

Also, after taking just a quick look at the changes here, it seems like this is something that could/should be done inside of a plugin (https://github.com/fastlane/fastlane/blob/master/fastlane/docs/Plugins.md) but I think we could have more discussion around that once I can better understand the use case.

Thanks! 🐙 🚀

**Comment by piotrdubiel** [comment]
Hey,
sorry for not elaborating more. In https://github.com/fastlane/fastlane/pull/81 there was extracting icon mentioned and also html template had commented place for app icon.
I am also using this exported icon to generate mail later, but main purpose of this PR is to use it in generated html.

File: fastlane/lib/fastlane/actions/s3.rb [hunk 1]
```diff
@@ -252,6 +264,51 @@ def self.expand_path_with_substitutions_from_ipa_plist(ipa, path)
...
+
+      def self.extract_icon_from_ipa(ipa_file, icon_file_name)
+        Zip::File.open(ipa_file) do |zipfile|
+          icon_file = self.largest_icon(zipfile.glob("**/Payload/**/#{icon_file_name}*.png"))
+          return nil unless icon_file
+          tmp_path = "/tmp/app_icon.png"
+          processed_tmp_path = "/tmp/processed_app_icon.png"
+          File.write(tmp_path, zipfile.read(icon_file))
+          # need to transform png from http://iphonedevwiki.net/index.php/CgBI_file_format
+          system("xcrun -sdk iphoneos pngcrush -revert-iphone-optimizations #{tmp_path} #{processed_tmp_path} >/dev/null")
```
Comment (ohwutup) [inline fastlane/lib/fastlane/actions/s3.rb]: Could you use the `Action.sh` method here?

Comment (piotrdubiel) [inline fastlane/lib/fastlane/actions/s3.rb]: Do you think it should print command/output or should it be silent?

Comment (ohwutup) [inline fastlane/lib/fastlane/actions/s3.rb]: I think it should probably print the command?

**Comment by KrauseFx** [comment]
@ohwutup seems like the PR was updated

File: fastlane/lib/fastlane/actions/s3.rb [hunk 2]
```diff
@@ -191,7 +196,14 @@ def self.run(config)
...
         ENV[SharedValues::S3_VERSION_OUTPUT_PATH.to_s] = version_url
 
+        Actions.lane_context[SharedValues::S3_ICON_OUTPUT_PATH] = icon_url
+        ENV[SharedValues::S3_ICON_OUTPUT_PATH.to_s] = icon_url
+
         UI.success("Successfully uploaded ipa file to '#{Actions.lane_context[SharedValues::S3_IPA_OUTPUT_PATH]}'")
+        UI.success("Successfully uploaded plist file to '#{Actions.lane_context[SharedValues::S3_PLIST_OUTPUT_PATH]}'")
+        UI.success("Successfully uploaded html file to '#{Actions.lane_context[SharedValues::S3_HTML_OUTPUT_PATH]}'")
+        UI.success("Successfully uploaded version file to '#{Actions.lane_context[SharedValues::S3_VERSION_OUTPUT_PATH]}'")
+        UI.success("Successfully uploaded icon file to '#{Actions.lane_context[SharedValues::S3_ICON_OUTPUT_PATH]}'")
```
Comment (ohwutup) [inline fastlane/lib/fastlane/actions/s3.rb]: These `Actions.lane_context[SharedValues::S3_*]` references are the same as the shorter `*_url` variables right? Might be nice to have these be the values that are interpolated here.

File: fastlane/lib/fastlane/actions/s3.rb [hunk 3]
```diff
@@ -252,6 +264,51 @@ def self.expand_path_with_substitutions_from_ipa_plist(ipa, path)
...
+          primary_icon = icons['CFBundlePrimaryIcon']
+          unless primary_icon.nil?
+            return primary_icon['CFBundleIconFiles']
+          end
+        end
+        nil
+      end
+
+      def self.upload_icon(ipa_file, info, url_part, bucket, acl)
+        icon_file = icon_files(info).first
```
Comment (ohwutup) [inline fastlane/lib/fastlane/actions/s3.rb]: Should we check that `icon_file` is not `nil` here before proceeding?

File: fastlane/lib/fastlane/actions/s3.rb [hunk 4]
```diff
@@ -252,6 +264,51 @@ def self.expand_path_with_substitutions_from_ipa_plist(ipa, path)
...
+        icon_file_basename = File.basename(icon_file)
+        icon_file_name = "#{url_part}#{icon_file_basename}.png"
+        icon_file_data = self.extract_icon_from_ipa(ipa_file, icon_file)
```
Comment (ohwutup) [inline fastlane/lib/fastlane/actions/s3.rb]: Also, if this fails, we should bail early too right?

**Comment by ohwutup** [comment]
Hey @piotrdubiel! Thanks for adding that context and fixing up the PR! I left a couple of little comments, feel free to disagree! Let me know what you think and we can get this pushed through soon! 👍

File: fastlane/lib/fastlane/actions/s3.rb (hunk 2 above)
Comment (piotrdubiel) [inline fastlane/lib/fastlane/actions/s3.rb]: I am not sure about this, I agree it's a little repetitive, but is explicit better in this case?

**Comment by piotrdubiel** [comment]
Hey @ohwutup I fixe

**Comment by piotrdubiel** [comment]
Hey, I fixed case when there's no icon in ipa. Also, internally I have `extract_app_icon` action separated from `s3` action, do you think it's a good idea?

**Comment by ohwutup** [comment]
Yeah! That would be great I think so people could run it on a one-off basis if necessary. Do you just call the `s3` action and then the `extract_app_icon` immediately after inside of your fastfile? 

Also, another option would be to separate them into their own actions, and then put an option in the `s3` action to decide whether or not the `extract_app_icon` action should also be run?

Curious to hear your thoughts!

**Comment by piotrdubiel** [comment]
To be precise I use it like this
```ruby
lane :upload do
    # stuff...
    extract_app_icon
    s3
end
```
`extract_app_icon` sets shared value `ICON_OUTPUT_PATH` and my modified `s3` action uses this as a default like this
```ruby
FastlaneCore::ConfigItem.new(key: :icon,
                             env_name: "",
                             description: "app icon file to upload",
                             optional: true,
                             default_value: Actions.lane_context[SharedValues::ICON_OUTPUT_PATH])
```
In this way is easy to use some custom icon if ipa doesn't contain one. What do you think?

**Comment by ohwutup** [comment]
Yeah! I think thats great. And I think that the `s3` action could conditionally call the `extract_app_icon` by passing an option to `s3` so that you can do it all in one line if you'd like too.

**Comment by ohwutup** [comment]
Hey @piotrdubiel, apologies for the delay here! Would you mind rebasing on master to fix any merge conflicts?

**Comment by ohwutup** [comment]
Hey @piotrdubiel ! I am going to close this issue as it has been pretty inactive for a while feel free to drop a note in here thought if you'd like us to reopen it once you're ready! Thanks again! 🚀

**Pull request closed without merging**